onnxconverter_common/auto_mixed_precision.py

```diff
@@ -0,0 +1,200 @@
+        i += outputs_per_iter
+    for out, dtype in out_to_dtype.items():
+        model.graph.value_info.append(helper.make_tensor_value_info(out, dtype, shape=None))
+    return model
+
+
+def get_tensor_values_using_ort(model, input_feed, output_names=None, sess_options=None):
+    if output_names is None:
+        sess = ort.InferenceSession(model.SerializeToString(), sess_options)
+        return sess.run(None, input_feed)
+    original_outputs = list(model.graph.output)
+    while len(model.graph.output) > 0:
+        model.graph.output.pop()
+    for n in output_names:
+        out = model.graph.output.add()
+        out.name = n
+    sess = ort.InferenceSession(model.SerializeToString(), sess_options)
+    try:
+        return sess.run(output_names, input_feed)
+    finally:
+        while len(model.graph.output) > 0:
+            model.graph.output.pop()
+        for orig_out in original_outputs:
+            out = model.graph.output.add()
+            out.CopyFrom(orig_out)
+
+
+class SegmentList:
+    def __init__(self, node_names):
+        self.node_names = node_names
+        self.first = NodeSegment(len(node_names))
+
+    def get_largest(self, adjacent_to_good=False):
+        adjacent_to_good = False
+        largest = None
+        current = self.first
+        prev_good = False
+        while current is not None:
+            can_use = not current.good and not current.bad
+            if adjacent_to_good:
+                next_good = current.next is not None and current.next.good
+                can_use = can_use and (prev_good or next_good)
+            if can_use and (largest is None or current.size > largest.size):
+                largest = current
+            prev_good = current.good
+            current = current.next
+        return largest
+
+    def get_nodes(self, node_segment=None):
+        i = 0
+        current = self.first
+        nodes = []
+        while current is not None:
+            if current is not node_segment and not current.good:
+                nodes.extend(self.node_names[i:i + current.size])
+            i += current.size
+            current = current.next
+        return nodes
+
+    def __repr__(self):
+        res = []
+        current = self.first
+        while current is not None:
+            res.append(current)
+            current = current.next
+        return repr(res)
+
+
+class NodeSegment:
+    def __init__(self, size):
+        self.size = size
+        self.next = None
+        self.good = False
+        self.bad = False
+
+    def split(self):
+        new_size = self.size // 2
+        new_segment = NodeSegment(self.size - new_size)
+        new_segment.next = self.next
+        self.next = new_segment
+        self.size = new_size
+
+    def __repr__(self):
+        if self.good:
+            return "*" + str(self.size) + "*"
+        if self.bad:
+            return "(" + str(self.size) + ")"
+        return str(self.size)
+
+
+if __name__ == "__main__":
+    main()
```

Comment:
Either make it a public API of the converter-common, or a standalone tool with argument.
The current main function implementation looks unusual.